---
title: "ST_5_model"
author: "Sean Swift"
date: "February 6, 2020"
output: 
  html_document:
        toc: true
        toc_depth: 4
        number_sections: true
---

#Description
ST_model focuses on modeling distributions of particular OTUs across the gradient.
This often requires subsetting by sample type to better understand changes in abundance.
Mostly making linear models. 

TODO:

  * Try Generalized Additive Models for Location, Scale and Shape (GAMLSS). See   https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2744-2 for comparison between GAMLSS and asin(sqrt())

  * Make lists of significant, abundant, variable otus and investigate them more thoroughly. 
  * Finally, with both (1) and this last bit, another useful thing is to relate effect (x) vs. abundance (y) or fit r2 (y) and either color by p or binary black/grey depending on p<0.05. You likely will have to transform the axes to get good spread and a nice “volcano”. But these will help you identify bugs that respond strongly to elevation (effect on x) and are either abundant, fit tightly to elevation, and/or are significant. Wikipedia the volcano plot for some examples, or just google image search it. 

  

# Pick out variable OTUs
Goal:
'“volcano plot” of -log10(p-value for shore_dist) vs. shore_dist slope or the relative abundance of OTUX to identify a few choice OTUs that are both abundant/change a lot and are highly significant'

## Load data

```{r setup}
#library(mvabund)
library(ggplot2)
library(dplyr)
library(microbiome)
library(vegan)

source("src/make_phyloseq.R")

ST_16S <- readRDS("data/processed/cleaned/ST_16S_spatial.rds")


# make phyloseq
ST_16S_phy <- make_phyloseq(ST_16S)


# get rid of plant shoots
# subset to just include terrestrial samples
# get rid of empty taxa

ST  <- subset_samples(ST_16S_phy, habitat != "Marine")
ST  <- subset_samples(ST, sample_type != "PlantShoot")
ST  <- prune_taxa(taxa_sums(ST) > 0, ST)



```


## Set up the models
To get started, define a function that generates linear models (lm) for each otu.


OTUs and values for X are pulled from a phyloseq object.
The model coefficients (intercept and slope) and summary coefficients (probablilities, errors, etc.) are stored in a data frame.

Might switch this over to manylm function in mvabund or gamlss for better (more complicated) model fitting.

### Z score models
This function generates linear models on the zscore of otu abundance as X and zscore of distance from shore as Y.

```{r zmodel, warning=TRUE}
# define a function to run linear models, output model coefficients, summary, and relative abundance
# function will also transform both dependent and independent variable to z-scores

get_z_models <- function(physeq, variable = "shore_dist") {
  ## Testing:
    variable = "shore_dist"
    physeq = ST_plant
  # 
  # pull out variable from sample data
  shore_dist <- physeq@sam_data[[variable]] %>% as.vector
  # transform to z score
  shore_dist_z  <- scale(shore_dist, center = T, scale = T)

  # pull out otu abundances

  physeq_rel <- transform_sample_counts(physeq, function(x) x / sum(x) )
  abund      <- otu_table(physeq) %>%  as.data.frame()
  rel_abund  <- otu_table(physeq_rel) %>% as.data.frame()
  asin_sqrt_abund <- apply(rel_abund, 1, function(x) asin(sqrt(x)))  %>% t() %>% as.data.frame()
  
  z_abund <- apply(asin_sqrt_abund, 1, function(x) scale(x)) %>% t() %>% as.data.frame()
  
  # apply model to each otu
  probs <- apply(z_abund, 1, function(x.row) {
          #TESTING: 
          x.row = z_abund[21,] %>% unlist()
          # *** MODEL DEFINITION ***
          mod1 <- lm(x.row ~ shore_dist)
          ####################################
          
          sum1 <- summary(mod1)
          
          append(sum1$coefficients, sum1$r.squared)
          
          
          }) %>%
          t() %>%
          as.data.frame()
  # set column names for model outputs
   colnames(probs) <-
  c(
  "intercept",
  "slope",
  "int_err",
  "slope_err",
  "int_t",
  "slope_t",
  "int_p",
  "slope_p",
  "r.squared"
  )
  # add otu names as a column
  probs$otu <- row.names(abund)
  # add otu relative abundance as a column
  probs$rel_abund <- rowSums(abund)/sum(rowSums(abund))
  
  # list otu abundance table and model information
  output <- list(probs,z_abund)
  names(output) <- c("models","model_data")
  return(output)
}

```

### arcsin square root model
This function generates linear models using the arcsin square root transformed otu relative abundance as X and distance from shore as Y.

```{r arcmodel}

# define a function to run linear models, output model coefficients, summary, and relative abundance
# function will also transform both dependent and independent variable to z-scores

get_arc_models <- function(physeq, variable = "shore_dist") {
  ## Testing:
  # variable = "shore_dist"
  # physeq = ST_plant
  # 
  # pull out variable from sample data
  shore_dist <- physeq@sam_data[[variable]] %>% as.vector
  
  # Pull out otu abundance and otu relative abundance (across samples)
  physeq_rel <- transform_sample_counts(physeq, function(x) x / sum(x) )
  abund      <- otu_table(physeq) %>%  as.data.frame()
  rel_abund  <- otu_table(physeq_rel) %>% as.data.frame()
  asin_sqrt_abund <- apply(rel_abund, 1, function(x) asin(sqrt(x)))  %>% t() %>% as.data.frame()
  # apply model to each otu
  probs <- apply(asin_sqrt_abund, 1, function(x.row) {
          #TESTING: 
          # x.row = abund[1200,] %>% unlist()
          # *** MODEL DEFINITION ***
          mod1 <- lm(x.row ~ shore_dist)
          ####################################
          
          sum1 <- summary(mod1)
          
          append(sum1$coefficients, sum1$r.squared)
          
          
          }) %>%
          t() %>%
          as.data.frame()
  # set column names for model outputs
  colnames(probs) <-
  c(
  "intercept",
  "slope",
  "int_err",
  "slope_err",
  "int_t",
  "slope_t",
  "int_p",
  "slope_p",
  "r.squared"
  )
  # add otu names as a column
  probs$otu <- row.names(abund)
  # add otu relative abundance as a column
  probs$rel_abund <- rowSums(abund)/sum(rowSums(abund))
  
  # add otu presence absence count as a column
  probs$PA_count <- decostand(abund, "pa") %>% rowSums
  
  # list otu abundance table and model information
  output <- list(probs,asin_sqrt_abund)
  names(output) <- c("models","model_data")
  return(output)
}

```

## Linear models : Look for abundant OTUs that change a lot
Having run the model, how many OTUs show up as significantly associated with distance from shore?
How many of those OTUs also vary in their abundance (slope of the linear regression)?
How many of those OTUs are also highly abundant?

To answer these questions, define functions for two types of "volcano" plots:

*1) p-value of OTU vs. variable slope (usually distance from shore)*
*2) p-value of OTU vs. relative abundance of OTU*

Plots include a vertical line at p = 0.001

```{r plots}
# Plot 1: P vs Slope
# define a function to plot p value of otu slope vs the actual slope coefficient
plot_pval_vs_slope <- function(got_models, description)
{
  plot(
    x = got_models$models$slope,
    y = -log10(got_models$models$slope_p),
    
    main = paste(description, ": P value of model vs. slope coefficient"),
    xlab = "shore_dist slope",
    ylab = "p value (-log10)",
  
    pch = 1,
    col = "darkblue"
    )
  abline(h = -log10(0.05))
}

# Plot 2: P vs Rel Abund
# define a function to plot p value of otu slope vs the relative abundance of the OTU
plot_pval_vs_RA <- function(got_models, description, physeq)
{
  plot(
    x = got_models$models$rel_abund,
    y = -log10(got_models$models$slope_p),
    
    main = paste(description, ": P value of model vs. relative abundance of OTU"),
    xlab = "relative abundance of OTU in data",
    ylab = "p value (-log10)",
    
    pch = 1,
    col = "darkred"
    )
  abline(h=-log10(0.05))
}
```

## Plot a regression for each OTU based on cutoff values
```{r}
# plot each OTU
model_view <- function(models_out = p_plants, physeq = ST_plant, p_cutoff = 0.001, presence_cutoff =0.5, description = "test", model_type = "arcsin(sqrt(RA)"){
  # for testing
   # models_out = p_plants
   # physeq = ST_plant
   # p_cutoff = 0.001
   # presence_cutoff =0.5
   # description = "test"
   # z = 1
  # get metadata
  sample_meta <-as(sample_data(physeq), "data.frame")
  sample_meta$id <- row.names(sample_meta)
  
  # subset by cutoffs
 cool_ps <- models_out$models %>% filter(slope_p < p_cutoff & PA_count > presence_cutoff * ncol(models_out$model_data))

 print(cool_ps$otu)

 # get the model info and plot each otu
 for(z in seq_along(cool_ps$otu)){
 
  # pull out OTU relative abundance
   otu_name <- cool_ps$otu[z]
   
   otu_mod <- models_out$models[otu_name,]
   
 an_otu <- models_out$model_data[otu_name,] %>% t() %>% as.data.frame()
 an_otu$id <- row.names(an_otu) 
 
 
  # for each OTU,  pull out metadata and taxonomy
 
 an_otu_tax <- as.data.frame(tax_table(ST_plant), stringsAsFactors = F)
 an_otu_tax <- an_otu_tax[otu_name,] 
 an_otu_tax <- paste(an_otu_tax[1,], collapse = ";")
 

 
 an_otu <- left_join(an_otu, sample_meta)
 colnames(an_otu)[1] <- "model_data"
 
 # plot relative abundance by the gradient
 
  p <-ggplot(an_otu,  aes( shore_dist, model_data, color = shore_dist)) +
    geom_point(size = 4) +
    labs(title = paste(cool_ps$otu[z], "Transformed abundance vs. gradient distance"),
         caption = paste("R^2 = ", round(otu_mod$r.squared, 3), "p =", otu_mod$slope_p)
         )+
    theme(title = element_text(size = 16, face = "bold"),
          axis.text=element_text(size=12),
          legend.text = element_text(size = 20),
          plot.subtitle = element_text(size = 12),
          plot.caption = element_text(size = 20))+
    labs(color = "Gradient Distance", x = "Gradient Distance", y = model_type,
          subtitle  = an_otu_tax) +
    scale_color_viridis_c(option = "C")+
    geom_abline(slope = otu_mod$slope, intercept = otu_mod$intercept)

 
  print(p)
  
  output_dir <- paste("outputs/figures/", description , "_otus/", sep = "")
  dir.create(output_dir)
  ggsave(paste(output_dir, description, "_", otu_name ,"_regression.png", sep = ""),
         width = 10, height = 7)
 }
 }
 
model_view()

```



### Linear models : All samples
Probably, running these models on all samples together will not be very informative, but let's try!

```{r allsamples}

# run linear model on all OTUs across all samples (phyloseq object "ST")
p_all_samples_z <- get_z_models(physeq = ST, variable = "shore_dist")
p_all_samples_arc <-get_arc_models(physeq = ST, variable = "shore_dist")


# how many otus have a tiny p value?
length(p_all_samples_z$models$slope_p[  p_all_samples_z$models$slope_p < 0.05  ] )

# plot 1
plot_pval_vs_slope(p_all_samples_z, "All samples, zscore")
plot_pval_vs_slope(p_all_samples_arc, "All samples, arcsin(sqrt)")


# plot 2
plot_pval_vs_RA(p_all_samples_z, "All samples, zscore")
plot_pval_vs_RA(p_all_samples_arc,"All samples, arcsin(sqrt)")
```

### Linear models : Subset Soil
Soil samples were collected at every transect site. Soil communities appear to vary linearly with the transect gradient.
Any good candidate OTUs varying across the gradient?

#### Soil

```{r soil}
# All soil samples

# Subset data to just soil samples
ST_soil <- subset_samples(ST, sample_type %in% c( "SoilShallow","SoilDeep"))
ST_soil <- prune_taxa(taxa_sums(ST_soil) > 0, ST_soil)

# run linear model on all OTUs across all samples (phyloseq object "ST")
p_soil_z   <- get_z_models(physeq = ST_soil, variable = "shore_dist")
p_soil_arc <- get_arc_models(physeq = ST_soil, variable = "shore_dist")
rm(ST_soil)

# how many otus have a tiny p value?
length(p_soil_z$models$slope_p[  p_soil_z$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_soil_z, "Soil")
plot_pval_vs_slope(p_soil_arc, "Soil")
# plot 2
plot_pval_vs_RA(p_soil_z, "Soil")
plot_pval_vs_RA(p_soil_arc, "Soil")


#####################################

# Shallow Soil Samples

# Subset data to just soil samples
ST_shal_soil <- subset_samples(ST, sample_type %in% c( "SoilShallow"))
ST_shal_soil <- prune_taxa(taxa_sums(ST_shal_soil) > 0, ST_shal_soil)

# run linear model on all OTUs across all samples (phyloseq object "ST")
p_shal_soil_z <- get_z_models(physeq = ST_shal_soil, variable = "shore_dist")
p_shal_soil_arc <- get_arc_models(physeq = ST_shal_soil, variable = "shore_dist")

# how many otus have a tiny p value?
length(p_shal_soil_z$models$slope_p[  p_shal_soil_z$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_shal_soil_z, "Shallow Soil, zscore")
plot_pval_vs_slope(p_shal_soil_arc, "Shallow Soil, arcsin(sqrt)")

# plot 2
plot_pval_vs_RA(p_shal_soil_z, "Shallow Soil, zscore")
plot_pval_vs_RA(p_shal_soil_arc, "Shallow Soil, arcsin(sqrt)")

# models for each OTU
model_view(models_out = p_shal_soil_arc, physeq = ST_shal_soil, description = "soil_shal")
```

### Linear models : Subset Sediment

```{r sediment}
# Sediment Samples

# Subset data to just soil samples
ST_sed <- subset_samples(ST, sample_type %in% c( "Sediment"))
ST_sed <- prune_taxa(taxa_sums(ST_sed) > 0, ST_sed)

# run linear model on all OTUs across all samples (phyloseq object "ST")
p_sed <- get_z_models(physeq = ST_sed, variable = "shore_dist")

# how many otus have a tiny p value?
length(p_sed$models$slope_p[  p_sed$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_sed, "Sediment")

# plot 2
plot_pval_vs_RA(p_sed, "Sediment")

```

### Linear models : Subset Only Water Samples
Water was collected at all transect sites.
Any good candidate OTUs varying across the gradient?

```{r water}
# Subset data to just plant root samples
ST_water <- subset_samples(ST, sample_type %in% c( "WaterNonSaline"))
ST_water <- prune_taxa(taxa_sums(ST_water) > 0, ST_water)

# run linear model on all OTUs across all samples (phyloseq object "ST")
p_water <- get_z_models(physeq = ST_water, variable = "shore_dist")
rm(ST_water)

# how many otus have a tiny p value?
length(p_water$models$slope_p[  p_water$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_water, "Water")

# plot 2
plot_pval_vs_RA(p_water, "Water")
```


### Linear models : Subset Only Plant Samples
Plant roots were collected at 3 sites per transect.
Any good candidate OTUs varying across the gradient?

```{r plant}
# Subset data to just plant root samples
ST_plant <- subset_samples(ST, sample_type %in% c( "PlantRoot"))
ST_plant <- prune_taxa(taxa_sums(ST_plant) > 0, ST_plant)

# pull out metadata
 plant_meta <- as(sample_data(ST_plant), "data.frame")
 plant_meta$id <- row.names(plant_meta)

# run linear model on all OTUs across all samples (phyloseq object "ST")
p_plants <- get_arc_models(physeq = ST_plant, variable = "shore_dist")


# how many otus have a tiny p value?
length(p_plant$models$slope_p[  p_plant$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_plant, "Plant Roots")

# plot 2
plot_pval_vs_RA(p_plant, "Plant Roots")



  # remove
 
 #rm(ST_plant)
```

### Linear models : Subset Only Mosquito Samples
Mosquitos were collected sort of haphazardly along the transects.
Any good candidate OTUs varying across the gradient?

```{r mosquito}
# Subset data to just mosquito samples
ST_mosquito <- subset_samples(ST, sample_type %in% c( "Mosquito"))
ST_mosquito <- prune_taxa(taxa_sums(ST_mosquito) > 0, ST_mosquito)
# run linear model on all OTUs across all samples (phyloseq object "ST")
p_mosquito <- get_z_models(physeq = ST_mosquito, variable = "shore_dist")
rm(ST_mosquito)

# how many otus have a tiny p value?
length(p_mosquito$models$slope_p[  p_mosquito$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_mosquito, "Mosquito")

# plot 2
plot_pval_vs_RA(p_mosquito, "Mosquito")
```


### Linear models : Subset Only Drosophila Samples
Drosophila were collected sort of haphazardly across the transects.
Any good candidate OTUs varying across the gradient?

```{r drosophila}
# Subset data to just drosophila samples
ST_drosophila <- subset_samples(ST, sample_type %in% c( "Drosophila"))
ST_drosophila <- prune_taxa(taxa_sums(ST_drosophila) > 0, ST_drosophila)
# run linear model on all OTUs across all samples (phyloseq object "ST")
p_drosophila <- get_z_models(physeq = ST_drosophila, variable = "shore_dist")
rm(ST_drosophila)

# how many otus have a tiny p value?
length(p_drosophila$models$slope_p[  p_drosophila$models$slope_p < 0.001  ] )

# plot 1
plot_pval_vs_slope(p_drosophila, "Drosophila")

# plot 2
plot_pval_vs_RA(p_drosophila, "Drosophila")
```

## Volcano plots
  Take the model outputs and make a volcano plot. Effect size (regression coefficient) will be on x axis.
  Y axis can be relative abundance, or fit. Color points by p value < 0.05
  
```{r}
# Make a basic volcano plot
  with(p_all_samples_z, plot(models$slope, models$slope_p, pch=20, main="Volcano plot"))

# Add colored points: red if models$slope_p<0.05, orange of log2FC>1, green if both)  
with(subset(p_all_samples_z$models, slope_p<.05 ), points(slope, slope_p, pch=20, col="red"))

```

