---
#title: "SuperTransect 16S"
output: html_notebook
---

```{r}
                        ###############################################################
                        #    ____                 ______                          __  #
                        #   / __/_ _____  ___ ___/_  __/______ ____  ___ ___ ____/ /_ #
                        #  _\ \/ // / _ \/ -_) __// / / __/ _ `/ _ \(_-</ -_) __/ __/ #
                        # /___/\_,_/ .__/\__/_/  /_/ /_/  \_,_/_//_/___/\__/\__/\__/  #
                        #         /_/                                                 #
                        ###############################################################
```

Welcome to the SuperTransect. 

This is Illumina Miseq data looking at 16S (V4 region, primers 515/806R)
Several sample types were sampled along 8 transects in Waimea Valley. 
Each sample was duplicated for DNA extraction and metabolomics. 

# Metadata Overview
Here's a quick breakdown of the metadata columns:

**sample_barcode** is the unique identifier for each sample.

These are all related to where the sample was collected. A little messy.

* **site_code**
* **transect_name**
* **site_name**
* **site_type**
* **distance** 

**lat** and **long** are in decimal degrees. Be careful, though, because some samples (like mosquitoes) have unique lat long decoupled from collection site.

These are all related to the sample type (e.g. plants vs. water)

* **sample_type**
* **host** 
* **habitat**
* **trophic** 

**analysis** indicated whether the sample was used for metabolomics (MS), dna extraction (DNA), or was split in the lab (MSandDNA).

**processing and notes** are mostly for sample management or any additional metadata (e.g. species names). Look at these if you need to go deeper.


```{r}
suppressWarnings(library(dplyr))
library(tidyr)
library(stringr)

file_metadata <- "../../data/raw/SuperTransect_All_Samples - all_collected_samples.csv"
metadata <- read.csv(file_metadata , header = T,colClasses = "character")
metadata <- metadata %>% select(-processing,-notes) %>% arrange(sample_id)

# Column names
metadata
```
### Transects

We sampled 8 transects in the Waimea watershed ranging from the the summit of the watershed, Kainapuaa (STSummit), to the beach of Waimea Bay (STEstuary).
```{r}
transects <- metadata %>% filter(site_type == "Transect") %>% arrange(lat)
unique(transects$transect_name) %>% as.character()

```

Each transect had 9 sampling points, spaced 100m apart. These can be accessed in the **distance** column.
The **site_code** column replicates this information. **site_name** is the unique combination of transect name and distance.

```{r}
# distances
unique(transects$distance) %>% as.character()

# site codes
unique(transects$site_code) %>% as.character()

# site names
unique(transects$site_name) %>% as.character()
```

### Marine Sites

The four marine sites were sampled the same week as the transects. However, they are not transects and do not have any 'distance' structure.
The sampling at these sites followed a bioblitz format. Host species are not necessarily the same across sites.
Because there is no distance, site name and site code are the same.


```{r}
marine <- metadata %>% filter(site_type == "Core Waimea")
# site name
unique(marine$site_name) %>% as.character()
# site code
unique(marine$site_code) %>% as.character()

```


# 16S Data Overview

Before we read in the whole dataset, let's look at the reads per step to see how read attrition looks in the pipeline.

```{r}
file_run_map <- "../../data/raw/st_miseq04_mapping_file - run_map.csv"
run_map <- read.csv(file_run_map, header = T, colClasses = "character")

file_read_step <- "../../data/interim/sequences_per_sample_per_step_97.tsv"
read_step <- read.table(file_read_step, sep = "\t",header = T)
read_step <- apply(read_step, 2, function(x) sub(" \\((.+) uniques\\)","",x)) %>% as.data.frame(stringsAsFactors=F)

colnames(read_step)[1] <- "id"

read_step

```

Let's also read in the mapping file

```{r}
read_step <- read_step %>% left_join(run_map, by = "id") %>% left_join(metadata, by = "sample_barcode")
read_step <- read_step %>% mutate(pass = ifelse(X12.Subsampling_97 == "/","No","Yes"))
read_step %>% count(pass, sample_type)

```

